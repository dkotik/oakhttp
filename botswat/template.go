package botswat

import (
	"errors"
	"fmt"
	"html/template"
	"strings"
	"time"

	_ "embed"

	"github.com/dkotik/oakhttp"
)

type templateOptions struct {
	Script         string
	Element        template.HTML
	Locale         string
	Title          string
	Description    string
	ErrorURL       string
	CookieName     string
	CookieDuration time.Duration
}

type TemplateOption func(*templateOptions) error

func NewTemplateOptions(all ...TemplateOption) (*templateOptions, error) {
	o := &templateOptions{}
	var err error
	for _, option := range append(all, func(o *templateOptions) (err error) {
		if o.Locale == "" {
			if err = WithLocale("en")(o); err != nil {
				return err
			}
		}
		if o.Title == "" {
			if err = WithTitle("Humanity Check")(o); err != nil {
				return err
			}
		}
		if o.Description == "" {
			if err = WithDescription("Please confirm your humanity to access this resource.")(o); err != nil {
				return err
			}
		}
		if o.ErrorURL == "" {
			if err = WithErrorURL("/")(o); err != nil {
				return err
			}
		}
		if o.CookieName == "" {
			if err = WithCookieName(DefaultCookieName)(o); err != nil {
				return err
			}
		}
		if o.CookieDuration == 0 {
			if err = WithCookieDuration(time.Minute * 5)(o); err != nil {
				return err
			}
		}
		return nil
	}) {
		if err = option(o); err != nil {
			return nil, fmt.Errorf("cannot initialize botswat template options: %w", err)
		}
	}
	return o, nil
}

func WithLocale(key string) TemplateOption {
	return func(o *templateOptions) error {
		if o.Locale != "" {
			return errors.New("site key is already set")
		}
		key = strings.TrimSpace(key)
		if key == "" {
			return errors.New("cannot use an empty site key")
		}
		o.Locale = key
		return nil
	}
}

func WithTitle(title string) TemplateOption {
	return func(o *templateOptions) error {
		if o.Title != "" {
			return errors.New("title is already set")
		}
		title = strings.TrimSpace(title)
		if title == "" {
			return errors.New("cannot use an empty title")
		}
		o.Title = title
		return nil
	}
}

func WithDescription(description string) TemplateOption {
	return func(o *templateOptions) error {
		if o.Description != "" {
			return errors.New("description is already set")
		}
		description = strings.TrimSpace(description)
		if description == "" {
			return errors.New("cannot use an empty description")
		}
		o.Description = description
		return nil
	}
}

func WithErrorURL(URL string) TemplateOption {
	return func(o *templateOptions) error {
		if o.ErrorURL != "" {
			return errors.New("error URL is already set")
		}
		URL = strings.TrimSpace(URL)
		if URL == "" {
			return errors.New("cannot use an empty error URL")
		}
		o.ErrorURL = URL
		return nil
	}
}

func WithCookieName(name string) TemplateOption {
	return func(o *templateOptions) error {
		if o.CookieName != "" {
			return errors.New("cookie name is already set")
		}
		name = strings.TrimSpace(name)
		if name == "" {
			return errors.New("cannot use an empty cookie name")
		}
		o.CookieName = name
		return nil
	}
}

func WithCookieDuration(d time.Duration) TemplateOption {
	return func(o *templateOptions) error {
		if o.CookieDuration != 0 {
			return errors.New("cookie duration is already set")
		}
		if d < time.Minute {
			return errors.New("cookie duration must be greater than 1 minute")
		}
		o.CookieDuration = d
		return nil
	}
}

func NewTemplate(withOptions ...TemplateOption) (*template.Template, error) {
	o := &templateOptions{}

	var err error
	defer func() {
		if err != nil {
			err = fmt.Errorf("cannot create botswat template: %w", err)
		}
	}()

	templateHTML, err := oakhttp.Templates.ReadFile("internal/templates/page/botswat.html")
	if err != nil {
		return nil, fmt.Errorf("unable to load the template file: %w", err)
	}

	for _, option := range append(withOptions, func(o *templateOptions) error {
		// if o.SiteKey == "" {
		// 	return errors.New("site key is required")
		// }
		return nil
	}) {
		if err = option(o); err != nil {
			return nil, err
		}
	}

	t, err := template.New("botswat").Parse(string(templateHTML))
	if err != nil {
		return nil, fmt.Errorf("cannot parse template: %w", err)
	}
	return t, nil
}
