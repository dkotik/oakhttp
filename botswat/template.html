<!DOCTYPE html>
<html lang="{{ block "locale" . }}en{{ end }}">
  <head>
    <meta charset="utf-8">
    <meta name="robots" content="noindex, follow">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <title>{{ .Title }}</title>
    <meta name="description" content="{{ .Description }}">
    <script src="{{ template "script" . }}" async defer></script>
    <style media="all">
      body {
        color: {{ block "textColor" .}}#333{{ end }};
        background-color: {{ block "backgroundColor" .}}#fff{{ end }};
        font-family: sans-serif;
        font-size: 1.4em;
      }

      article {
        margin: 4em auto;
        max-width: 25em;
      }

      a {
        display: block;
        text-align: center;
        text-transform: capitalize;
        text-decoration: none;
        padding: 1em 1.5em;
        border-radius: 0.5em;
        color: {{ template "backgroundColor" . }};
        background-color: {{ block "actionColor" .}}#0edd7a{{ end }};
      }

      {{ template "style" . }}
    </style>
    <script type="text/javascript">
    function toggle(ID, on) {
      var element = document.getElementById(ID);
      element.style.display = on ? "block" : "none";
    }

    function callback(token) {
      const d = new Date();
      // d.setTime(d.getTime() + (exdays*24*60*60*1000));
      d.setTime(d.getTime() + ({{ .CookieDuration.Seconds }}*1000));
      let expires = "expires="+ d.toUTCString();
      document.cookie = {{ .CookieName }} + "=" + token + ";" + expires + ";path=/";


      toggle("success", true);
      toggle("error", false);
      // console.log("new token", token);
      // setTimeout(function(){window.location.reload();}, 2000);
    }

    function callbackError(err) {
      toggle("success", false);
      toggle("error", true);
    }
    </script>
  </head>
  <body>
    <article>
      <h1>{{ .Title }}</h1>
      <p>{{ .Description }}</p>

      {{ template "element" . }}

      <!-- Requires enabled Javascript and Cookies to complete. -->

      <section id="success" style="display:none;">
        <p>You have proven your humanity!</p>
        <a href="#proceed" onclick="window.location.reload()">Proceed</a>
      </section>

      <section id="error" style="display:none;">
        <p>
          Having trouble reaching humanity validation server...
        </p>
        <a href="#back" onclick="window.location.reload()">Refresh</a>
      </section>
    </article>
  </body>
</html>
