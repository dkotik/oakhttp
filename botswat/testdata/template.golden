<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8">
    <meta name="robots" content="noindex, follow">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <title></title>
    <meta name="description" content="">
    <script src="" async defer></script>
    <style media="all">
      body {
        color: #333;
        background-color: #fff;
        font-family: sans-serif;
        font-size: 1.4em;
      }

      article {
        margin: 4em auto;
        max-width: 25em;
      }

      a {
        display: block;
        text-align: center;
        text-transform: capitalize;
        text-decoration: none;
        padding: 1em 1.5em;
        border-radius: 0.5em;
        color: #fff;
        background-color: #0edd7a;
      }
    </style>
    <script type="text/javascript">
    function toggle(ID, on) {
      var element = document.getElementById(ID);
      element.style.display = on ? "block" : "none";
    }

    function callback(token) {
      const d = new Date();
      
      d.setTime(d.getTime() + ( 0 *1000));
      let expires = "expires="+ d.toUTCString();
      document.cookie = "" + "=" + token + ";" + expires + ";path=/";


      toggle("success", true);
      toggle("error", false);
      
      
    }

    function callbackError(err) {
      toggle("success", false);
      toggle("error", true);
    }
    </script>
  </head>
  <body>
    <article>
      <h1></h1>
      <p></p>

      

      

      <section id="success" style="display:none;">
        <p>You have proven your humanity!</p>
        <a href="#proceed" onclick="window.location.reload()">Proceed</a>
      </section>

      <section id="error" style="display:none;">
        <p>
          Having trouble reaching humanity validation server...
        </p>
        <a href="#back" onclick="window.location.reload()">Refresh</a>
      </section>
    </article>
  </body>
</html>
